File: vm/cpp/src/CPU.cpp
#include <iostream>
#include "../header/CPU.h"
#include "../header/utils.hpp"
#include "../header/State.h"


CPU::CPU(Bus* bus)
{
    this->_bus = bus;
}

void CPU::step() const
{
    this->_bus->state->accept(this->_bus->state->fetch());
    this->_bus->state->eval();
}

void CPU::halt(uint8_t code)
{
    cout << red 
    << "cpu_.halt 0x" 
    << u32hex(reinterpret_cast<uint32_t*>(code), sizeof(code)) 
    << white << endl;
}

File: vm/rust/src/cpu.rs
#[derive(Debug)]
pub struct CPU {
    pub state: State,
    pub bus: Bus,
}

impl CPU {
    pub fn new(&b: Bus) -> Bus
    {
        CPU { bus: &b, state: &b.state }
    }

    pub fn step(&self)
    {
        self.state.accept(self.state.fetch());
        self.state.eval();
    }
}